File: lib/core/crypto/bse2.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:flutter_rust_bridge/flutter_rust_bridge.dart'
    show AnyhowException;

import '../rust/rust_facade.dart';

/// The end-to-end-encrypted container for cloud transfer links — the same
/// "BSE2" format the web client writes and reads.
///
/// The cryptography lives in ONE place: `bishare-protocol/src/bse2.rs` (the
/// canonical spec, pinned to WebCrypto golden vectors). This class is a thin
/// Dart face over that Rust code via flutter_rust_bridge, so the app seals
/// its Remote Share uploads and opens web uploads with byte-identical logic —
/// and with hardware AES where the platform has it, off the Dart event loop.
///
/// Container format ("BSE2" v1):
///   header (24 bytes):
///     [0..4)   magic  = "BSE2"
///     [4]      version = 1
///     [5..8)   reserved (0)
///     [8..12)  salt (4 random bytes, folded into every nonce)
///     [12..16) recordSize (uint32 BE, plaintext bytes per record)
///     [16..24) plaintextSize (uint64 BE)
///   then back-to-back records, record i:
///     AES-256-GCM(plaintext[i*RS ...], iv = salt ‖ u64BE(i), aad = u32BE(i))
///     — each record's ciphertext is its plaintext length + 16-byte tag.
///
/// The 32-byte key rides in the share link's URL fragment (`#k=<base64url>`),
/// which browsers never send to the server — so the relay only ever stores
/// ciphertext, and whoever holds the link (not the code alone) can open it.
class Bse2Exception implements Exception {
  const Bse2Exception(this.message);
  final String message;

  @override
  String toString() => 'Bse2Exception: $message';
}

abstract final class Bse2 {
  static const List<int> magic = [0x42, 0x53, 0x45, 0x32]; // "BSE2"
  static const int version = 1;
  static const int headerSize = 24;
  static const int tagSize = 16;
  static const int keySize = 32;

  /// Plaintext bytes per record — what every production writer uses.
  static const int recordSize = 1024 * 1024;

  /// Fresh random 32-byte key for one transfer (Rust CSPRNG).
  static Uint8List generateKey() => Rust.bse2GenerateKey();

  /// URL-safe base64, no padding — the text after `#k=` in the share link.
  static String encodeKey(List<int> raw) =>
      base64Url.encode(raw).replaceAll('=', '');

  /// Decode a `#k=` fragment value to the raw 32-byte key, or null if it
  /// isn't one. Tolerates padding, like the web and Rust readers.
  static Uint8List? decodeKey(String s) {
    final b64 = s.replaceAll('-', '+').replaceAll('_', '/');
    final padded = b64.padRight(b64.length + (4 - b64.length % 4) % 4, '=');
    try {
      final raw = base64.decode(padded);
      return raw.length == keySize ? Uint8List.fromList(raw) : null;
    } on FormatException {
      return null;
    }
  }

  /// Exact container length for a plaintext of [plaintextSize] bytes. The
  /// uploader must reserve THIS, not the plaintext size — the relay only ever
  /// sees ciphertext. Mirrors `bse2::ciphertext_size` (asserted equal in
  /// tests); pure Dart so size gates work before the bridge is touched.
  static int ciphertextSize(int plaintextSize) {
    final records = plaintextSize == 0
        ? 1
        : (plaintextSize + recordSize - 1) ~/ recordSize;
    return headerSize + plaintextSize + records * tagSize;
  }

  /// True when [file] starts with the BSE2 magic — i.e. it is an encrypted
  /// container, not the plain payload.
  static Future<bool> sniff(File file) async {
    final raf = await file.open();
    try {
      final head = await raf.read(magic.length);
      if (head.length < magic.length) return false;
      for (var i = 0; i < magic.length; i++) {
        if (head[i] != magic[i]) return false;
      }
      return true;
    } finally {
      await raf.close();
    }
  }

  /// Seal [input] into a new container at [output] under [key] (a fresh
  /// random salt is chosen in Rust). [onProgress] reports plaintext bytes
  /// done / total. Throws [Bse2Exception] on any failure; the partial output
  /// is removed by Rust, so a caller can never upload a stub.
  static Future<void> encryptFile({
    required File input,
    required File output,
    required Uint8List key,
    void Function(int plainBytes, int plainTotal)? onProgress,
  }) async {
    if (key.length != keySize) {
      throw const Bse2Exception('key must be 32 bytes');
    }
    try {
      await for (final p in Rust.bse2EncryptFile(
        inputPath: input.path,
        outputPath: output.path,
        key: key,
      )) {
        onProgress?.call(p.done.toInt(), p.total.toInt());
      }
    } on Object catch (e) {
      throw Bse2Exception(_message(e));
    }
  }

  /// Open the container at [input] into the plaintext file at [output].
  /// Throws [Bse2Exception] on a bad key, tampered/reordered records,
  /// truncation, trailing bytes or an unsupported header — never partial
  /// output reported as success. [onProgress] reports plaintext bytes
  /// produced / total.
  static Future<void> decryptFile({
    required File input,
    required File output,
    required Uint8List key,
    void Function(int plainBytes, int plainTotal)? onProgress,
  }) async {
    if (key.length != keySize) {
      throw const Bse2Exception('key must be 32 bytes');
    }
    try {
      await for (final p in Rust.bse2DecryptFile(
        inputPath: input.path,
        outputPath: output.path,
        key: key,
      )) {
        onProgress?.call(p.done.toInt(), p.total.toInt());
      }
    } on Object catch (e) {
      throw Bse2Exception(_message(e));
    }
  }

  /// Rust delivers failures into the progress stream as a plain message (the
  /// FRB sink-fn rule); unwrap whatever wrapper the bridge put around it.
  static String _message(Object e) {
    if (e is Bse2Exception) return e.message;
    if (e is AnyhowException) return e.message;
    return e.toString();
  }
}

File: rust/src/api/bse2.rs
//! BSE2 — the end-to-end-encrypted cloud-link container
//! (`bishare_protocol::bse2`), exposed to Dart as whole-file operations so the
//! app's Remote Share uploads ciphertext exactly like the web client does, and
//! decrypts web uploads with the same shared code. Crypto runs in Rust (hardware
//! AES where the platform has it), never on the Dart event loop.

use std::fs::{self, File};
use std::io::{BufReader, BufWriter};

use bishare_protocol::bse2;

use crate::frb_generated::StreamSink;

/// Progress of a whole-file seal/open: plaintext bytes done out of total.
pub struct Bse2Progress {
    pub done: u64,
    pub total: u64,
}

/// Fresh random 32-byte AES-256 key for one transfer.
#[flutter_rust_bridge::frb(sync)]
pub fn bse2_generate_key() -> Vec<u8> {
    bse2::generate_key().to_vec()
}

/// URL-safe base64 (no padding) — the text after `#k=` in a share link.
#[flutter_rust_bridge::frb(sync)]
pub fn bse2_encode_key(key: Vec<u8>) -> Option<String> {
    let k: [u8; bse2::KEY_SIZE] = key.try_into().ok()?;
    Some(bse2::encode_key(&k))
}

/// Inverse of [`bse2_encode_key`]; `None` unless it decodes to 32 bytes.
#[flutter_rust_bridge::frb(sync)]
pub fn bse2_decode_key(fragment: String) -> Option<Vec<u8>> {
    bse2::decode_key(&fragment).map(|k| k.to_vec())
}

/// Exact container length for a plaintext of this size — what the uploader
/// must reserve, since the relay only ever sees ciphertext.
#[flutter_rust_bridge::frb(sync)]
pub fn bse2_ciphertext_size(plaintext_size: u64) -> u64 {
    bse2::ciphertext_size(plaintext_size)
}

/// True when `head` (the first bytes of a file) carries the container magic.
#[flutter_rust_bridge::frb(sync)]
pub fn bse2_is_container(head: Vec<u8>) -> bool {
    bse2::is_container(&head)
}

/// Seal the file at `input_path` into a new container at `output_path` with a
/// fresh random salt. Runs on its own thread and returns at once; progress
/// streams through `sink`, and the stream closing without an error means the
/// output is complete and synced to disk. A failure is delivered INTO the
/// sink via `add_error` (FRB sink-fn rule — see `quic_send_file`), and the
/// partial output file is removed so a caller can never upload a stub.
pub fn bse2_encrypt_file(
    sink: StreamSink<Bse2Progress>,
    input_path: String,
    output_path: String,
    key: Vec<u8>,
) {
    std::thread::spawn(move || {
        if let Err(e) = encrypt_inner(&sink, &input_path, &output_path, &key) {
            let _ = fs::remove_file(&output_path);
            let _ = sink.add_error(e);
        }
    });
}

fn encrypt_inner(
    sink: &StreamSink<Bse2Progress>,
    input_path: &str,
    output_path: &str,
    key: &[u8],
) -> Result<(), String> {
    let k: [u8; bse2::KEY_SIZE] = key
        .try_into()
        .map_err(|_| bse2::Bse2Error::InvalidKey.to_string())?;
    let input = File::open(input_path).map_err(|e| format!("open input: {e}"))?;
    let total = input
        .metadata()
        .map_err(|e| format!("stat input: {e}"))?
        .len();
    let output = File::create(output_path).map_err(|e| format!("create output: {e}"))?;
    let mut writer = BufWriter::new(output);
    bse2::encrypt(
        BufReader::new(input),
        &mut writer,
        &k,
        bse2::generate_salt(),
        bse2::RECORD_SIZE,
        total,
        |done| {
            let _ = sink.add(Bse2Progress { done, total });
        },
    )
    .map_err(|e| e.to_string())?;
    // Everything on disk before Dart opens the file for upload.
    writer
        .into_inner()
        .map_err(|e| format!("flush output: {e}"))?
        .sync_all()
        .map_err(|e| format!("sync output: {e}"))?;
    Ok(())
}

/// Open the container at `input_path` into the plaintext file at
/// `output_path`. Same threading/error contract as [`bse2_encrypt_file`]; on
/// any failure (wrong key, tampering, truncation) the partial output is
/// removed — the caller never sees half a file reported as success.
pub fn bse2_decrypt_file(
    sink: StreamSink<Bse2Progress>,
    input_path: String,
    output_path: String,
    key: Vec<u8>,
) {
    std::thread::spawn(move || {
        if let Err(e) = decrypt_inner(&sink, &input_path, &output_path, &key) {
            let _ = fs::remove_file(&output_path);
            let _ = sink.add_error(e);
        }
    });
}

fn decrypt_inner(
    sink: &StreamSink<Bse2Progress>,
    input_path: &str,
    output_path: &str,
    key: &[u8],
) -> Result<(), String> {
    let k: [u8; bse2::KEY_SIZE] = key
        .try_into()
        .map_err(|_| bse2::Bse2Error::InvalidKey.to_string())?;
    let input = File::open(input_path).map_err(|e| format!("open input: {e}"))?;
    let output = File::create(output_path).map_err(|e| format!("create output: {e}"))?;
    let mut writer = BufWriter::new(output);
    bse2::decrypt(BufReader::new(input), &mut writer, &k, |done, total| {
        let _ = sink.add(Bse2Progress { done, total });
    })
    .map_err(|e| e.to_string())?;
    writer
        .into_inner()
        .map_err(|e| format!("flush output: {e}"))?
        .sync_all()
        .map_err(|e| format!("sync output: {e}"))?;
    Ok(())
}
